types.go

```diff
@@ -140,36 +146,35 @@ func Is(any Any, v interface{}) bool {
 // returned verbatim. If it is of type proto.Message, it will be marshaled as a
 // protocol buffer. Otherwise, the object will be marshaled to json.
 func MarshalAny(v interface{}) (Any, error) {
-	var marshal func(v interface{}) ([]byte, error)
-	switch t := v.(type) {
-	case Any:
-		// avoid reserializing the type if we have an any.
+	if t, ok := v.(Any); ok {
 		return t, nil
-	case proto.Message:
-		marshal = func(v interface{}) ([]byte, error) {
-			return proto.Marshal(t)
-		}
-	case gogoproto.Message:
-		marshal = func(v interface{}) ([]byte, error) {
-			return gogoproto.Marshal(t)
-		}
-	default:
-		marshal = json.Marshal
 	}
 
-	url, err := TypeURL(v)
+	url, isJSON, err := typeURL(v)
 	if err != nil {
 		return nil, err
 	}
-
-	data, err := marshal(v)
+	a := anyType{
+		typeURL: url,
+	}
+	if !isJSON {
+		switch t := v.(type) {
+		case proto.Message:
+			a.value, err = proto.Marshal(t)
+		case gogoproto.Message:
+			a.value, err = gogoproto.Marshal(t)
```

Comment:
This one should probably only be fore the v2.2 branch, or if this patch is not super important, we could just drop this (and don't have this patch for v2.2.x?)